File: jianzhioffer/firstround/jianzhi_36.py
# Definition for a Node.
class Node(object):
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def __init__(self):
        self.head = None
        self.pre = None

    def treeToDoublyList(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        def dfs(node):
            if not node:
                return
            dfs(node.left)
            if self.pre:
                self.pre.right = node
                node.left = self.pre
            else:
                self.head = node
            self.pre = node
            dfs(node.right)

        if not root:
            return
        dfs(root)
        self.head.left = self.pre
        self.pre.right = self.head
        return self.head


if __name__ == '__main__':
    l1 = Node(4)
    l2 = Node(2)
    l3 = Node(5)
    l4 = Node(1)
    l5 = Node(3)

    l1.left = l2
    l1.right = l3
    l2.left = l4
    l2.right = l5

    obj = Solution()
    result = obj.treeToDoublyList(l1)
    print(result.val, result.left.val)

File: jianzhioffer/secondround/jianzhi_36.cpp
#include <cstddef>
using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};

//中序遍历，时间复杂度O(N)，空间复杂度O(1)
class Solution {
public:
    Node* treeToDoublyList(Node* root) {
        if(root==NULL) return NULL;
        prev = NULL;
        DFS(root);
        head->left = prev;
        prev->right = head;
        return head;    
    }

    void DFS(Node* node)
    {
        if(node==NULL) return;
        DFS(node->left);
        if (prev!=NULL)
        {
            prev->right = node;
            node->left = prev;
        }
        else
        {
            head = node;
        }
        prev = node;
        DFS(node->right);
        
    }

private:
    Node* head;
    Node* prev;
};